# mini_proj_02_20171541_차영민

In [1]:
import sys
import sklearn
import numpy as np
import os
import pandas as pd
from sklearn.utils import shuffle

## train_csv 데이터 받아오기
###### train_csv를 받아온 후, shuffle을 통해 데이터를 섞는다. (subject로 나열 되어 재대로된 값이 나오지 않을 수 있기 때문에)

In [2]:
data_train = pd.read_csv("train.csv")
data_train = shuffle(data_train, random_state = 42)
data_train

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
4525           0.283203          -0.047024          -0.168986   
1446           0.256904          -0.036623          -0.133856   
5995           0.291316          -0.001065          -0.072461   
4222           0.276116          -0.010909          -0.102886   
6754           0.256382           0.000428          -0.113664   
...                 ...                ...                ...   
5191           0.278897          -0.030306          -0.096043   
5226           0.289183          -0.049248          -0.125083   
5390           0.293946          -0.018341          -0.119916   
860            0.280475          -0.018976          -0.113756   
7270           0.263582           0.006928          -0.095320   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
4525          0.384949          0.176898         -0.310332          0.381757   
1446          0.201409         -0.154142          0.344183          0.092771   
5995         -0.336609         -0.279162         -0.303323         -0.381421   
4222         -0.992196         -0.982169         -0.981127         -0.992838   
6754          0.075014          0.046502         -0.369482         -0.024794   
...                ...               ...               ...               ...   
5191         -0.555352         -0.104055         -0.438064         -0.572530   
5226         -0.290043         -0.212102         -0.469731         -0.307317   
5390         -0.627198         -0.216566         -0.424764         -0.648666   
860          -0.994825         -0.985314         -0.965857         -0.995170   
7270         -0.368655         -0.142631         -0.151250         -0.426026   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
4525          0.122611         -0.332984          0.465563  ...   
1446         -0.182114          0.291899          0.586004  ...   
5995         -0.253026         -0.337381          0.074430  ...   
4222         -0.981366         -0.979864         -0.931924  ...   
6754         -0.085337         -0.385018          0.562230  ...   
...                ...               ...               ...  ...   
5191         -0.112149         -0.429688         -0.479744  ...   
5226         -0.209558         -0.528635         -0.283244  ...   
5390         -0.253814         -0.417569         -0.547082  ...   
860          -0.984285         -0.963293         -0.937503  ...   
7270         -0.130656         -0.149079         -0.216841  ...   

      fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
4525                        -0.932568                    -0.034924   
1446                        -0.338095                     0.017986   
5995                        -0.698254                    -0.044663   
4222                        -0.906098                    -0.177275   
6754                        -0.316113                     0.347143   
...                               ...                          ...   
5191                         0.200267                    -0.155279   
5226                        -0.360982                    -0.019533   
5390                         0.348736                    -0.414621   
860                         -0.984051                    -0.079515   
7270                        -0.158376                     0.217989   

      angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
4525                              0.558036                          0.258975   
1446                             -0.475545                          0.942947   
5995                             -0.551806                         -0.680774   
4222                             -0.309304                          0.176515   
6754                              0.590475                          0.831084   
...                                    ...                               ...   
5191                             -0.137981                

In [3]:
data_train.describe()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
count        7352.000000        7352.000000        7352.000000   
mean            0.274488          -0.017695          -0.109141   
std             0.070261           0.040811           0.056635   
min            -1.000000          -1.000000          -1.000000   
25%             0.262975          -0.024863          -0.120993   
50%             0.277193          -0.017219          -0.108676   
75%             0.288461          -0.010783          -0.097794   
max             1.000000           1.000000           1.000000   

       tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
count       7352.000000       7352.000000       7352.000000       7352.000000   
mean          -0.605438         -0.510938         -0.604754         -0.630512   
std            0.448734          0.502645          0.418687          0.424073   
min           -1.000000         -0.999873         -1.000000         -1.000000   
25%           -0.992754         -0.978129         -0.980233         -0.993591   
50%           -0.946196         -0.851897         -0.859365         -0.950709   
75%           -0.242813         -0.034231         -0.262415         -0.292680   
max            1.000000          0.916238          1.000000          1.000000   

       tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
count       7352.000000       7352.000000       7352.000000  ...   
mean          -0.526907         -0.606150         -0.468604  ...   
std            0.485942          0.414122          0.544547  ...   
min           -1.000000         -1.000000         -1.000000  ...   
25%           -0.978162         -0.980251         -0.936219  ...   
50%           -0.857328         -0.857143         -0.881637  ...   
75%           -0.066701         -0.265671         -0.017129  ...   
max            0.967664          1.000000          1.000000  ...   

       fBodyBodyGyroJerkMag-skewness()  fBodyBodyGyroJerkMag-kurtosis()  \
count                      7352.000000                      7352.000000   
mean                         -0.307009                        -0.625294   
std                           0.321011                         0.307584   
min                          -0.995357                        -0.999765   
25%                          -0.542602                        -0.845573   
50%                          -0.343685                        -0.711692   
75%                          -0.126979                        -0.503878   
max                           0.989538                         0.956845   

       angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
count                  7352.000000                           7352.000000   
mean                      0.008684                              0.002186   
std                       0.336787                              0.448306   
min                      -0.976580                             -1.000000   
25%                      -0.121527                             -0.289549   
50%                       0.009509                              0.008943   
75%                       0.150865                              0.292861   
max                       1.000000                              1.000000   

       angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
count                       7352.000000                           7352.000000   
mean                           0.008726                             -0.005981   
std                            0.608303                              0.477975   
min                           -1.000000                             -1.000000   
25%                           -0.482273                             -0.376341   
50%                            0.008735                             -0.000368   
75%                            0.506187                              0.359368   
max                            0.998702                     

In [4]:
data_train.shape

(7352, 563)

###### data개수 : 7352개, 특성 개수 : 563개(레이블 포함)

## x_train과 y_train 선언

### 선언 후, x_train과 y_train이 잘 나왔는지 확인

In [5]:
x_train = data_train.drop(["Activity"], axis = 1)
y_train = data_train["Activity"]
x_train

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
4525           0.283203          -0.047024          -0.168986   
1446           0.256904          -0.036623          -0.133856   
5995           0.291316          -0.001065          -0.072461   
4222           0.276116          -0.010909          -0.102886   
6754           0.256382           0.000428          -0.113664   
...                 ...                ...                ...   
5191           0.278897          -0.030306          -0.096043   
5226           0.289183          -0.049248          -0.125083   
5390           0.293946          -0.018341          -0.119916   
860            0.280475          -0.018976          -0.113756   
7270           0.263582           0.006928          -0.095320   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
4525          0.384949          0.176898         -0.310332          0.381757   
1446          0.201409         -0.154142          0.344183          0.092771   
5995         -0.336609         -0.279162         -0.303323         -0.381421   
4222         -0.992196         -0.982169         -0.981127         -0.992838   
6754          0.075014          0.046502         -0.369482         -0.024794   
...                ...               ...               ...               ...   
5191         -0.555352         -0.104055         -0.438064         -0.572530   
5226         -0.290043         -0.212102         -0.469731         -0.307317   
5390         -0.627198         -0.216566         -0.424764         -0.648666   
860          -0.994825         -0.985314         -0.965857         -0.995170   
7270         -0.368655         -0.142631         -0.151250         -0.426026   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
4525          0.122611         -0.332984          0.465563  ...   
1446         -0.182114          0.291899          0.586004  ...   
5995         -0.253026         -0.337381          0.074430  ...   
4222         -0.981366         -0.979864         -0.931924  ...   
6754         -0.085337         -0.385018          0.562230  ...   
...                ...               ...               ...  ...   
5191         -0.112149         -0.429688         -0.479744  ...   
5226         -0.209558         -0.528635         -0.283244  ...   
5390         -0.253814         -0.417569         -0.547082  ...   
860          -0.984285         -0.963293         -0.937503  ...   
7270         -0.130656         -0.149079         -0.216841  ...   

      fBodyBodyGyroJerkMag-skewness()  fBodyBodyGyroJerkMag-kurtosis()  \
4525                        -0.730593                        -0.932568   
1446                        -0.032958                        -0.338095   
5995                        -0.351997                        -0.698254   
4222                        -0.623358                        -0.906098   
6754                         0.020630                        -0.316113   
...                               ...                              ...   
5191                         0.417731                         0.200267   
5226                        -0.021970                        -0.360982   
5390                         0.527152                         0.348736   
860                         -0.889671                        -0.984051   
7270                         0.124435                        -0.158376   

      angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
4525                    -0.034924                              0.558036   
1446                     0.017986                             -0.475545   
5995                    -0.044663                             -0.551806   
4222                    -0.177275                             -0.309304   
6754                     0.347143                              0.590475   
...                           ...                                   ...   
5191                    -0.155279            

In [6]:
y_train

4525    WALKING_DOWNSTAIRS
1446    WALKING_DOWNSTAIRS
5995               WALKING
4222               SITTING
6754    WALKING_DOWNSTAIRS
               ...        
5191               WALKING
5226    WALKING_DOWNSTAIRS
5390               WALKING
860               STANDING
7270               WALKING
Name: Activity, Length: 7352, dtype: object

## Pipeline 및 StandardScaler사용

In [7]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

num_pipline = Pipeline([
    ('std_scaler',StandardScaler()),
    ('Imputer', SimpleImputer(strategy="median"))
])
x_train = num_pipline.fit_transform(x_train)

In [8]:
from sklearn.svm import SVC
from sklearn.ensemble import ExtraTreesClassifier, AdaBoostClassifier, VotingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import confusion_matrix

## SVC(C 값을 조기종료 방식으로 찾기)

### cv10이란 함수 선언(cv10을 이용해 비교하여 더 내려갈 때, C값 찾기)

In [9]:
def cv10(model, x_train, y_train):
    cv10_val_score = cross_val_score(model,x_train,y_train, scoring="accuracy", cv=10)
    cv10_score = np.mean(cv10_val_score)
    return cv10_score

In [10]:
svm_clf = SVC(kernel = 'rbf', random_state=42)

best_cv10_score = 0
error_going_up = 0
best_C = 0
for C in range(10,1000,1):
    C = C / 10;
    svm_clf.C = C
    svm_clf.fit(x_train, y_train)
    scores_checking = cv10(svm_clf,x_train, y_train)
    if scores_checking > best_cv10_score:
        best_cv10_score = scores_checking
        error_going_up = 0
        print(f"cv : {best_cv10_score}, C: {C}")
        best_C = C
    else:
        error_going_up += 1
        print(f"over the best_cv!! {scores_checking} C: {C}, best_C: {best_C} ,error_count: {error_going_up}")
        if error_going_up == 30:
            svm_clf.C = best_C
            break

cv : 0.9786451863354036, C: 1.0
cv : 0.9789171103223898, C: 1.1
cv : 0.979325273587696, C: 1.2
cv : 0.9804133392487431, C: 1.3
cv : 0.9809573720792665, C: 1.4
cv : 0.9813657202011239, C: 1.5
cv : 0.9816376441881101, C: 1.6
cv : 0.9820456225968648, C: 1.7
cv : 0.9823179162969536, C: 1.8
cv : 0.982590025140491, C: 1.9
cv : 0.9831340579710146, C: 2.0
cv : 0.9832701123927832, C: 2.1
over the best_cv!! 0.9832701123927832 C: 2.2, best_C: 2.1 ,error_count: 1
cv : 0.9835422212363205, C: 2.3
cv : 0.9836780908015379, C: 2.4
cv : 0.9839500147885241, C: 2.5
cv : 0.9843583629103817, C: 2.6
cv : 0.9847667110322391, C: 2.7
cv : 0.9849025805974565, C: 2.8
cv : 0.9850384501626739, C: 2.9
cv : 0.9857183525584146, C: 3.0
over the best_cv!! 0.985582298136646 C: 3.1, best_C: 3.0 ,error_count: 1
cv : 0.9861265158237208, C: 3.2
cv : 0.9862625702454896, C: 3.3
cv : 0.9865344942324755, C: 3.4
cv : 0.986806603076013, C: 3.5
over the best_cv!! 0.986806603076013 C: 3.6, best_C: 3.5 ,error_count: 1
over the best_c

over the best_cv!! 0.9891182342502217 C: 14.4, best_C: 12.6 ,error_count: 18
over the best_cv!! 0.9891182342502217 C: 14.5, best_C: 12.6 ,error_count: 19
over the best_cv!! 0.9889821798284532 C: 14.6, best_C: 12.6 ,error_count: 20
over the best_cv!! 0.9889821798284532 C: 14.7, best_C: 12.6 ,error_count: 21
over the best_cv!! 0.9889821798284532 C: 14.8, best_C: 12.6 ,error_count: 22
over the best_cv!! 0.9889821798284532 C: 14.9, best_C: 12.6 ,error_count: 23
over the best_cv!! 0.9888461254066844 C: 15.0, best_C: 12.6 ,error_count: 24
over the best_cv!! 0.9888459405501331 C: 15.1, best_C: 12.6 ,error_count: 25
over the best_cv!! 0.9888459405501331 C: 15.2, best_C: 12.6 ,error_count: 26
over the best_cv!! 0.9888459405501331 C: 15.3, best_C: 12.6 ,error_count: 27
over the best_cv!! 0.9888459405501331 C: 15.4, best_C: 12.6 ,error_count: 28
over the best_cv!! 0.9888459405501331 C: 15.5, best_C: 12.6 ,error_count: 29
over the best_cv!! 0.9888459405501331 C: 15.6, best_C: 12.6 ,error_count: 30

In [11]:
best_C

12.6

In [12]:
svm_best_clf = SVC(kernel = 'rbf', C = best_C, random_state=42, probability = True)
svm_best_clf.fit(x_train, y_train)

SVC(C=12.6, probability=True, random_state=42)

## Adaboost (base_estimator = DecisionTreeClassifier)

###### Adaboosting 역시 조기종료방식(n_estimators을 찾는 방식)을 사용했으나, Desktop에서 24시간 넘게 돌려도 안끝나서 계산된 것 중 최상의 값을 넣었습니다.

In [13]:
ada_best_clf = AdaBoostClassifier(base_estimator = DecisionTreeClassifier(max_depth=10, 
                                                                      min_samples_split=3, random_state=42),
                              random_state=42, n_estimators = 160)

## 각각의 모델들의 K-fold의 평균 값 출력

In [14]:
model = (svm_best_clf,ada_best_clf)
for x in model:
    test = cv10(x,x_train,y_train)
    print(f"model : {x}, cv: {test}")

model : SVC(C=12.6, probability=True, random_state=42), cv: 0.9892542886719905
model : AdaBoostClassifier(base_estimator=DecisionTreeClassifier(max_depth=10,
                                                         min_samples_split=3,
                                                         random_state=42),
                   n_estimators=160, random_state=42), cv: 0.990342539189589


* SVC : 0.9892542886719905
* Ada : 0.990342539189589

## Voting(SVC, Ada)

#### Voting 방식 : soft

In [15]:
vot_clf_best = VotingClassifier(
    estimators=[('svm', svm_best_clf), ('ada',ada_best_clf)], 
    voting = 'soft', n_jobs = -1)
vot_clf_best.fit(x_train, y_train)

VotingClassifier(estimators=[('svm',
                              SVC(C=12.6, probability=True, random_state=42)),
                             ('ada',
                              AdaBoostClassifier(base_estimator=DecisionTreeClassifier(max_depth=10,
                                                                                       min_samples_split=3,
                                                                                       random_state=42),
                                                 n_estimators=160,
                                                 random_state=42))],
                 n_jobs=-1, voting='soft')

## cv10 이후 값 출력

In [16]:
score_best = cross_val_score(vot_clf_best, x_train, y_train, cv=10, scoring="accuracy")
score_best

array([0.99320652, 0.99320652, 0.99319728, 0.98639456, 0.98911565,
       0.99455782, 0.99455782, 0.99455782, 0.99591837, 0.99455782])

In [17]:
np.mean(score_best)

0.9929270186335405

### Voting시 가장 잘 된 값 ->SVC, Ada
##### Logistic, Extratrees는 뺄 때 더 높아서 SVC, Ada로 Voting.

# mean : 0.9929270186335405